File: CEOI/2001/ceoionline2001/Day2/select/P/query.pas
Unit Query;
Interface
  Function Size:Integer;
  Function Member(i,j:Integer):Integer;
  Procedure Answer(i:Integer);
Implementation
Const
  MaxN=30000;
  MaxL=16;
  Init:Boolean=True;
  B:Array[Boolean] of 0..1=(0,1);
Var
  N,
  M,
  Half,
  FullS,
  NQ:Integer;       {# queries}
  T:Array[0..MaxN] of Integer;
  D:Array[0..MaxN] of Integer;
  OutF:Text;
  Bit:Set of 0..15 Absolute M;

 Procedure EndQuery;
 Begin
   Assign(OutF, 'select.out'); Rewrite(OutF);
   Writeln(OutF, '0 Point'); Flush(OutF);
   Close(OutF);
   Halt;
  End{EndQuery};

  Procedure Start;
  Var i:Integer;
    InFile:Text;
  Begin
    Assign(OutF, 'select.out'); Rewrite(OutF);
    Assign(InFile, 'select.in'); Reset(InFile);
    ReadLn(InFile, N);
    Close(InFile);

    For i:=1 To N Do Begin
      T[i]:=-1;
      D[i]:=0;
    End;
    T[0]:=0; D[0]:=0;
    If Odd(N) Then M:=N Else M:=N-1;
    Half:=M div 2 +1;

    Init:=False;
    NQ:=0;
    FullS:=0;
    For i:=0 To MaxL-1 Do
      If i In Bit Then Inc(FullS);
  End{Start};

  Function Find(x:Integer):Integer;
  Var Nx,xx:Integer;
  Begin
    Nx:=x;
    While T[Nx]>0 Do Nx:=T[Nx];
    While x<>Nx Do Begin{path compression}
      xx:=T[x]; T[x]:=Nx;
      x:=xx;
    End;
    Find:=Nx;
  End{Find};

  Procedure Union(Var Nx:Integer; Ny:Integer);
  Var X:Integer;
  Begin
    If Nx=0 Then
      Nx:=Ny
    Else If Ny<>0 Then Begin
      If T[Nx]>T[Ny] Then Begin
        X:=Nx; Nx:=Ny; Ny:=X;
      End;
      T[Nx]:=T[Nx]+T[Ny];
      T[Ny]:=Nx;
    End;
  End{Union};

  Function Size:Integer;
  Begin
    If Init Then Start;
    WriteLn(OutF, 'Size= ',N); Flush(OutF);
    Size:=N;
  End{Size};

Function Member(i,j:Integer):Integer;
Var Ri,Rj,Di,Dj,Si,Sj:Integer;
  Ui,Vi,Uj,Vj:Integer;
  SDi,SDj:Integer;
  X:Word;
  SX:Set of 0..15 Absolute X;
  Ans:Boolean;
Begin
  If Init Then Begin
    Writeln(OutF, 'ERROR: Illegal dialog. Size must be called first!');
    EndQuery;
  End;
  write(OutF, i:1,' ',j,' ');
  If (i<1)Or(i>N)Or (j<1)Or(j>N) Then Begin
    writeln(OutF, i:1,' ',j:1);
    writeln(OutF, 'ERROR: Value out of range'); Flush(OutF);
    EndQuery;
  End;

  Inc(NQ);
  Ans:=False;
  Ri:=Find(i);
  Rj:=Find(j);
  Di:=D[Ri]; Dj:=D[Rj];
  If Ri=Rj Then Begin
    writeln(OutF, 1);Flush(OutF);
    Member:=1;
    Exit;
  End;
  If (Ri=Dj)Or(Rj=Di) Then Begin
    writeln(OutF, 0);Flush(OutF);
    Member:=0;
    Exit;
  End;
  Si:=Abs(T[Ri]); Sj:=Abs(T[Rj]);
  SDi:=Abs(T[Di]); SDj:=Abs(T[Dj]);
  If Si>=SDi Then Begin
    Ui:=Ri; Vi:=Di;
  End Else Begin
    Ui:=Di; Vi:=Ri;
    X:=Si; Si:=SDi; SDi:=X
  End;
  If Sj>=SDj Then Begin
    Uj:=Rj; Vj:=Dj;
  End Else Begin
    Uj:=Dj; Vj:=Rj;
    X:=Sj; Sj:=SDj; SDj:=X
  End;
  If Si+Sj>=Half Then Begin
    Union(Ui, Vj);
    Union(Vi, Uj);
    D[Ui]:=Vi; If Vi<>0 Then D[Vi]:=Ui;
    Ans:=(Find(i)=Find(j));
  End Else Begin
    If Si+Sj<=SDi+SDj+2 Then Begin
      Union(Ui, Uj);
      Union(Vi, Vj);
      D[Ui]:=Vi; If Vi<>0 Then D[Vi]:=Ui;
      Ans:=(Find(i)=Find(j));
    End Else Begin
      If Si+SDj=Sj+SDi Then Begin
        X:=Si+SDj+Sj+SDi;
        If SX * Bit=SX Then begin
          Union(Ui, Vj);
          Union(Vi, Uj);
          D[Ui]:=Vi; If Vi<>0 Then D[Vi]:=Ui;
          Bit:=Bit-SX;
          Dec(Half, Si+SDj);
	  Ans:=(Find(i)=Find(j));
        End Else Begin
          Union(Ui, Uj);
          Union(Vi, Vj);
          D[Ui]:=Vi; If Vi<>0 Then D[Vi]:=Ui;
	  Ans:=(Find(i)=Find(j));
        End;
      End Else begin{Si+SDj<>Sj+SDi}
        Union(Ui, Vj);
        Union(Vi, Uj);
        D[Ui]:=Vi; If Vi<>0 Then D[Vi]:=Ui;
	 Ans:=(Find(i)=Find(j));
      End;
    End;
  End;

  Member:=B[Ans];
  writeln(OutF,B[Ans]);  Flush(OutF);
End{Member};

Procedure Answer(i:Integer);
Var Ri, Di,x,y,Nx:Integer;
  Sum:Integer;
  OK:Boolean;
  G:Array[0..MaxN] Of Boolean;
  a,b:Integer;
Begin
  If Init Then Begin
    Writeln(OutF, 'ERROR: Illegal dialog. Size must be called first!');
    EndQuery;
  End;
  If (i<1)Or(i>N) Then Begin
    writeln(OutF, 'Your Answer: ',i);
    writeln(OutF, 'ERROR: Value out of range');
    EndQuery;
  End;

  Ri:=Find(i);
  Di:=D[Ri];
  Sum:=0;

  For x:=1 To N Do Begin
    Nx:=Find(x);
    G[x]:= (Nx<>Ri)And(
           (T[Nx]<T[D[Nx]])Or
           (T[Nx]=T[D[Nx]]) And (Nx<D[Nx])Or
           (Nx=Di)
                      );
  End;
  For x:=1 To N Do
    If (T[x]<0)And(x<>Di)And(x<>Ri) Then Begin
      y:=D[x];
      If T[x]<T[y] Then
        Sum:=Sum+Abs(T[x])
      Else
        Sum:=Sum+Abs(T[y]);
      T[x]:=0; T[y]:=0;
    End;
  {for x=1..N};
  Sum:=Sum+Abs(T[Di]);
  OK:=Sum<N Div 2+1;

  If OK Then
    WriteLn(OutF, 'Your Answer: ',i, ', Correct')
  Else
    WriteLn(OutF, 'Your Answer: ',i, ', Not Correct');

  If OK Then
    WriteLn(OutF, 'Majority Group:')
  Else
    WriteLn(OutF, 'Non-Majority Group:');
  a:=0; b:=0; G[0]:=True; G[N+1]:=True;
  For x:=1 To N+1 Do Begin
    If Not G[x] Then Begin
      If Not G[x-1] Then
        b:=x
      Else Begin
        a:=x; b:=a;
      End;
    End Else Begin
      If a+1<b Then
        Write(OutF, a:1,'..',b:1,' ')
      Else If a<b Then
        Write(OutF, a:1,' ',b:1,' ')
      Else If (b=a)And(a<>0) Then
        Write(OutF, a:1,' ');
      b:=0;
    End;
  End; WriteLn(OutF);
  If OK Then
    WriteLn(OutF,'Non-Majority Group:')
  Else
    WriteLn(OutF,'Majority Group:');
  a:=0; b:=0; G[0]:=False; G[N+1]:=False;
  For x:=1 To N+1 Do Begin
    If G[x] Then Begin
      If G[x-1] Then
        b:=x
      Else Begin
        a:=x; b:=a;
      End;
    End Else Begin
      If a+1<b Then
        Write(OutF, a:1,'..',b:1,' ')
      Else If a<b Then
        Write(OutF, a:1,' ',b:1,' ')
      Else If (b=a)And(a<>0) Then
        Write(OutF, a:1,' ');
      b:=0;
    End;
  End; WriteLn(OutF);

  WriteLn(OutF, 'Number of Queries: ',NQ:1,' ');

  If OK Then Begin
    Sum:=N-NQ; If Sum<0 Then Sum:=0;
    WriteLn(OutF,'Full Possible Score: ', FullS);
    WriteLn(OutF,'Your Score: ',Sum:1);
  End Else Begin
    WriteLn(OutF,'Full Possible Score: ', FullS);
    WriteLn(OutF,'Your Score: ',0:1);
  End;
 Close(OutF);
End{Answer};

End.

File: CEOI/2001/ceoionline2001/Day2/select/C/query.c
#include "query.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MaxN            30000
#define MaxL            16
#define true 1
#define false 0

static int Init = true;
static char B[2];
static int FullS, NQ;   /*# queries*/
static unsigned int N, M, Half;
static int T[MaxN + 1];
static int D[MaxN + 1];
static FILE *OutF;

static void EndQuery(void)
{
  OutF = fopen("select.out", "w");
  fprintf(OutF, "0 Point\n");
  fflush(OutF);
  if (OutF != NULL)
    fclose(OutF);
  OutF = NULL;
  exit(0);
}

static void Start(void)
{
  int i;
  FILE *InFile;
  int FORLIM;

  B[0] = 0; 
  B[1] = 1;
  InFile = NULL;
  OutF = fopen("select.out", "w");
  InFile = fopen("select.in", "r");
  fscanf(InFile, "%hd%*[^\n]", &N);
  getc(InFile);
  if (InFile != NULL)
    fclose(InFile);
  InFile = NULL;

  FORLIM = N;
  for (i = 1; i <= FORLIM; i++) {
    T[i] = -1;
    D[i] = 0;
  }
  T[0] = 0;
  D[0] = 0;
  if (N & 1)
    M = N;
  else
    M = N - 1;
  Half = M / 2 + 1;

  Init = false;
  NQ = 0;
  FullS = 0;
  for (i = 0; i < MaxL; i++) {
    if ( ((1 << i) & M) != 0 )
      FullS++;
  }
  if (InFile != NULL)
    fclose(InFile);
}

static int Find(int x)
{
  int Nx, xx;

  Nx = x;
  while (T[Nx] > 0)
    Nx = T[Nx];
  while (x != Nx) {  /*path compression*/
    xx = T[x];
    T[x] = Nx;
    x = xx;
  }
  return Nx;
}

static void Union(int *Nx, int Ny)
{
  int X;

  if (*Nx == 0) {
    *Nx = Ny;
    return;
  }
  if (Ny == 0)
    return;
  if (T[*Nx] > T[Ny]) {
    X = *Nx;
    *Nx = Ny;
    Ny = X;
  }
  T[*Nx] += T[Ny];
  T[Ny] = *Nx;
}

int Size(void)
{
  if (Init)
    Start();
  fprintf(OutF, "Size= %d\n", N);
  fflush(OutF);
  return N;
}

int Member(int i, int j)
{
  int Result, Ri, Rj, Di, Dj, Si, Sj, Ui, Vi, Uj, Vj, SDi, SDj;
  unsigned int X;
  int Ans;

  if (Init) {
    fprintf(OutF, "ERROR: Illegal dialog. Size must be called first!\n");
    EndQuery();
  }
  fprintf(OutF, "Member(%d, %d)", i, j);
  if (i < 1 || i > N || j < 1 || j > N) {
    fprintf(OutF, "Member(%d, %d)\n", i, j);
    fprintf(OutF, "ERROR: Value out of range\n");
    fflush(OutF);
    EndQuery();
  }

  NQ++;
  Ans = false;
  Ri = Find(i);
  Rj = Find(j);
  Di = D[Ri];
  Dj = D[Rj];
  if (Ri == Rj) {
    fprintf(OutF, "%d\n", 1);
    fflush(OutF);
    return 1;
  }
  if (Ri == Dj || Rj == Di) {
    fprintf(OutF, "%d\n", 0);
    fflush(OutF);
    return 0;
  }
  Si = abs(T[Ri]);
  Sj = abs(T[Rj]);
  SDi = abs(T[Di]);
  SDj = abs(T[Dj]);
  if (Si >= SDi) {
    Ui = Ri;
    Vi = Di;
  } else {
    Ui = Di;
    Vi = Ri;
    X = Si;
    Si = SDi;
    SDi = X;
  }
  if (Sj >= SDj) {
    Uj = Rj;
    Vj = Dj;
  } else {
    Uj = Dj;
    Vj = Rj;
    X = Sj;
    Sj = SDj;
    SDj = X;
  }
  if (Si + Sj >= Half) {
    Union(&Ui, Vj);
    Union(&Vi, Uj);
    D[Ui] = Vi;
    if (Vi != 0)
      D[Vi] = Ui;
    Ans = (Find(i)==Find(j));
  } else {
    if (Si + Sj <= SDi + SDj + 2) {
      Union(&Ui, Uj);
      Union(&Vi, Vj);
      D[Ui] = Vi;
      if (Vi != 0)
	D[Vi] = Ui;
      Ans = (Find(i)==Find(j));
    } else {
      if (Si + SDj == Sj + SDi) {
	X = Si + SDj + Sj + SDi;
	if ((X & M) == X) {
	  Union(&Ui, Vj);
	  Union(&Vi, Uj);
	  D[Ui] = Vi;
	  if (Vi != 0)
	    D[Vi] = Ui;
	    M &= ~X;
	  Half += -Si - SDj;
          Ans = (Find(i)==Find(j));
	} else {
	  Union(&Ui, Uj);
	  Union(&Vi, Vj);
	  D[Ui] = Vi;
	  if (Vi != 0)
	    D[Vi] = Ui;
         Ans = (Find(i)==Find(j));
	}
      } else {  /*Si+SDj<>Sj+SDi*/
	Union(&Ui, Vj);
	Union(&Vi, Uj);
	D[Ui] = Vi;
	if (Vi != 0)
	  D[Vi] = Ui;
       Ans = (Find(i)==Find(j));
      }
    }
  }

  Result = B[Ans];
  fprintf(OutF, "=%d\n", B[Ans]);
  fflush(OutF);
  return Result;
}

void Answer(int i)
{
  int Ri, Di, x, y, Nx, Sum;
  int OK;
  int G[MaxN + 1];
  int a, b, FORLIM;

  if (Init) {
    fprintf(OutF, "ERROR: Illegal dialog. Size must be called first!\n");
    EndQuery();
  }
  if (i < 1 || i > N) {
    fprintf(OutF, "Your Answer: %d\n", i);
    fprintf(OutF, "ERROR: Value out of range\n");
    EndQuery();
  }

  Ri = Find(i);
  Di = D[Ri];
  Sum = 0;

  FORLIM = N;
  for (x = 1; x <= FORLIM; x++) {
    Nx = Find(x);
    G[x] = (Nx != Ri && (T[Nx] < T[D[Nx]] ||
			 T[Nx] == T[D[Nx]] && Nx < D[Nx] || Nx == Di));
  }
  FORLIM = N;
  for (x = 1; x <= FORLIM; x++) {
    if (T[x] < 0 && x != Di && x != Ri) {
      y = D[x];
      if (T[x] < T[y])
	Sum += abs(T[x]);
      else
	Sum += abs(T[y]);
      T[x] = 0;
      T[y] = 0;
    }
  }
  /*for x=1..N*/
  Sum += abs(T[Di]);
  OK = (Sum <= N / 2);
  fprintf(OutF, "Your Answer: %d", i);
  if (OK) {
    fprintf(OutF, ", Correct\n");
    } else {
    fprintf(OutF, ", Not Correct\n");
  }

  if (OK)
    fprintf(OutF, "Majority Group:\n");
  else
    fprintf(OutF, "Non-Majority Group:\n");
  a = 0;
  b = 0;
  G[0] = true;
  G[N + 1] = true;
  FORLIM = N + 1;
  for (x = 1; x <= FORLIM; x++) {
    if (!G[x]) {
      if (!G[x - 1])
	b = x;
      else {
	a = x;
	b = a;
      }
    } else {
      if (a + 1 < b)
	fprintf(OutF, "%d..%d ", a, b);
      else if (a < b)
	fprintf(OutF, "%d %d ", a, b);
      else if (b == a && a != 0)
	fprintf(OutF, "%d ", a);
      b = 0;
    }
  }
  putc('\n', OutF);
  if (OK)
    fprintf(OutF, "Non-Majority Group:\n");
  else
    fprintf(OutF, "Majority Group:\n");
  a = 0;
  b = 0;
  G[0] = false;
  G[N + 1] = false;
  FORLIM = N + 1;
  for (x = 1; x <= FORLIM; x++) {
    if (G[x]) {
      if (G[x - 1])
	b = x;
      else {
	a = x;
	b = a;
      }
    } else {
      if (a + 1 < b)
	fprintf(OutF, "%d..%d ", a, b);
      else if (a < b)
	fprintf(OutF, "%d %d ", a, b);
      else if (b == a && a != 0)
	fprintf(OutF, "%d ", a);
      b = 0;
    }
  }
  fprintf(OutF, "\nNumber of Queries: %d \n", NQ);

  if (OK) {
    Sum = N - NQ;
    if (Sum < 0)
      Sum = 0;
    fprintf(OutF, "Full Possible Score: %d\n", FullS);
    fprintf(OutF, "Your Score: %d\n", Sum);
  } else {
    fprintf(OutF, "Full Possible Score: %d\n", FullS);
    fprintf(OutF, "Your Score: %d\n", 0);
  }
  if (OutF != NULL)
    fclose(OutF);
  OutF = NULL;
}

